### PR TITLE
add points UI

### PR DESCRIPTION
**Pull request by jeremy-babylonlabs**

### CODE REVIEW

**Comment by jeremy-babylonlabs** [comment]
Screenshot 2024-09-06 at 3 32 28 PM
Screenshot 2024-09-06 at 3 32 22 PM
Screenshot 2024-09-06 at 3 32 16 PM
Screenshot 2024-09-06 at 3 32 03 PM

major changes:
- moved `publicKeyNoCoord`, `address`, `btcWallet`, `btcWalletNetwork` states from pages.tsx to contexts, so that the api calls can be called more atomicly
- added points UI for both desktop and mobile

File: src/app/api/getDelegationPoints.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import { Pagination } from "../types/api";
+
+import { apiWrapper } from "./apiWrapper";
+
+export interface DelegationPoint {
+  staking_tx_hash_hex: string;
+  last_updated_btc_height: number;
+  staker: {
+    pk: string;
+    polkadot: number;
```
Comment (vitsalis) [inline src/app/api/getDelegationPoints.ts:12]: We should not use `polkadot` in the API, this is a name reserved for a different crypto. We should use `points` everywhere cc @jrwbabylonlab

File: src/app/api/getDelegationPoints.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  data: DelegationPoint[];
+  pagination: Pagination;
+}
+
+export const getDelegationPoints = async (
+  key: string,
+  stakerBtcPk?: string,
+): Promise<PaginatedDelegationPoints> => {
+  if (!stakerBtcPk) {
+    throw new Error("Staker_btc_pk must be provided");
```
Comment (vitsalis) [inline src/app/api/getDelegationPoints.ts:30]:
```suggestion
    throw new Error("Staker BTC PK must be provided");
```

File: src/app/api/getDelegationPoints.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  }
+
+  const params: Record<string, string> = {
+    pagination_key: encode(key),
+  };
+
+  if (stakerBtcPk) {
```
Comment (vitsalis) [inline src/app/api/getDelegationPoints.ts:37]: We have already checked before that this is set so we don't need this `if` condition. We can directly set `staker_btc_pk` above.

File: src/app/api/getDelegationPoints.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
+import { encode } from "url-safe-base64";
+
+import { Pagination } from "../types/api";
+
+import { apiWrapper } from "./apiWrapper";
+
+export interface DelegationPoint {
```
Comment (vitsalis) [inline src/app/api/getDelegationPoints.ts:7]:
```suggestion
export interface DelegationPoints {
```

File: src/app/api/getDelegationPoints.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    pk: string;
+    polkadot: number;
+  };
+  finality_provider: {
+    pk: string;
+    polkadot: number;
+  };
+}
+
+export interface PaginatedDelegationPoints {
```
Comment (vitsalis) [inline src/app/api/getDelegationPoints.ts:20]:
```suggestion
export interface PaginatedDelegationsPoints {
```

File: src/app/api/getStakerPoints.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import { encode } from "url-safe-base64";
+
+import { apiWrapper } from "./apiWrapper";
+
+export interface StakerPoint {
```
Comment (vitsalis) [inline src/app/api/getStakerPoints.ts:5]:
```suggestion
export interface StakerPoints {
```

File: src/app/api/getStakerPoints.ts [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  staker_btc_pk: string;
+  polkadots: number;
+}
+
+export const getStakerPoints = async (
```
Comment (vitsalis) [inline src/app/api/getStakerPoints.ts:10]:
```suggestion
export const getStakersPoints = async (
```

As we can either pass a single one or an array.

File: src/app/api/getStakerPoints.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+  stakerBtcPk: string | string[],
```
Comment (vitsalis) [inline src/app/api/getStakerPoints.ts:11]: Why don't we make this to only receive an array of strings to make our life easier and have a unified way to call it?

File: src/app/api/getStakerPoints.ts [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+  const response = await apiWrapper(
+    "GET",
+    "/v1/points/stakers",
+    "Error getting staker points",
+    params,
+  );
+
+  return response.data;
+};
```
Comment (vitsalis) [inline src/app/api/getStakerPoints.ts:29]: Commenting here as I can't comment on the `src/app/assets/coins.svg` image: Let's rename the file to `src/app/assets/points.svg` to avoid any form of confusion.

File: src/app/components/Delegations/Delegation.tsx [hunk 10]
```diff
@@ -1,9 +1,11 @@
+import Image from "next/image";
 import { useEffect, useState } from "react";
 import { AiOutlineInfoCircle } from "react-icons/ai";
 import { FaBitcoin } from "react-icons/fa";
 import { IoIosWarning } from "react-icons/io";
 import { Tooltip } from "react-tooltip";
 
+import coins from "@/app/assets/coins.svg";
```
Comment (vitsalis) [inline src/app/components/Delegations/Delegation.tsx:8]:
```suggestion
import points from "@/app/assets/points.svg";
```

File: src/app/components/Delegations/Delegation.tsx [hunk 11]
```diff
@@ -162,6 +164,20 @@ export const Delegation: React.FC<DelegationProps> = ({
             <Tooltip id={`tooltip-${stakingTxHash}`} className="tooltip-wrap" />
           </div>
         </div>
+        <div className="relative flex justify-end lg:justify-center order-5">
+          <div className="flex items-center justify-end gap-1">
+            <Image
+              src={coins}
+              alt="Coins"
```
Comment (vitsalis) [inline src/app/components/Delegations/Delegation.tsx:171]:
```suggestion
              alt="Points"
```

File: src/app/components/Delegations/Delegation.tsx [hunk 12]
```diff
@@ -162,6 +164,20 @@ export const Delegation: React.FC<DelegationProps> = ({
...
+              width={16}
+              height={16}
+              className="text-gray-600"
+            />
+            <p className="whitespace-nowrap font-semibold">
+              {points ? maxDecimals(points, 8) : 0}
```
Comment (vitsalis) [inline src/app/components/Delegations/Delegation.tsx:177]: What number of decimals are we expecting for points? cc @jrwbabylonlab

File: src/app/components/Delegations/Delegations.tsx [hunk 13]
```diff
@@ -113,7 +146,7 @@ export const Delegations: React.FC<DelegationsProps> = ({
         id,
         delegationsAPI,
         publicKeyNoCoord,
-        btcWalletNetwork,
+        btcWalletNetwork!,
```
Comment (vitsalis) [inline src/app/components/Delegations/Delegations.tsx:149]: Why is this update needed?

File: src/app/context/btcWallet/BtcWallet.tsx [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+  };
+
+  return (
+    <BtcWalletContext.Provider value={value}>
+      {children}
+    </BtcWalletContext.Provider>
+  );
+};
+
+export const useBtcWallet = () => {
```
Comment (vitsalis) [inline src/app/context/btcWallet/BtcWallet.tsx:50]: The introduction of this seems unrelated to the PR. Why did we add it here instead of a separate PR that can be reviewed independently?

File: src/app/components/Summary/Summary.tsx [hunk 15]
```diff
@@ -46,70 +59,92 @@ export const Summary: React.FC<SummaryProps> = ({
...
-            {typeof btcWalletBalanceSat === "number" ? (
+          <div className="divider xl:divider-horizontal xl:mx-4 my-0" />
+          <div className="flex flex-1 flex-col gap-2 text-sm xl:items-center">
+            <div className="flex items-center gap-1">
+              <p className="dark:text-neutral-content">Total points</p>
+            </div>
+            <div className="flex items-center gap-1">
+              <Image
+                src={coins}
+                alt="Coins"
```
Comment (vitsalis) [inline src/app/components/Summary/Summary.tsx:97]:
```suggestion
                alt="Points"
```

File: src/app/context/btcWallet/BtcWallet.tsx [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+import React, { createContext, ReactNode, useContext, useState } from "react";
+
+import { WalletProvider } from "@/utils/wallet/wallet_provider";
+interface BtcWalletContextType {
+  publicKeyNoCoord: string;
+  address: string;
+  btcWallet: WalletProvider | undefined;
+  btcWalletNetwork: networks.Network | undefined;
+  setPublicKeyNoCoord: (key: string) => void;
+  setAddress: (address: string) => void;
```
Comment (vitsalis) [inline src/app/context/btcWallet/BtcWallet.tsx:11]: Why do we need the `setPublicKeyNoCoord`, `setAddress` and `setBtcWalletNetwork` methods? These data should not be different than what `btcWallet` returns.

**Comment by jeremy-babylonlabs** [comment]
i think i misclicked it on mobile

**Review by totraev — DISMISSED** [review]

File: src/app/api/getDelegationPoints.ts [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+export interface PaginatedDelegationsPoints {
+  data: DelegationPoints[];
+  pagination: Pagination;
+}
+
+export const getDelegationPoints = async (
+  paginationKey?: string,
+  stakerBtcPk?: string,
+  stakingTxHashHexes?: string[],
+): Promise<PaginatedDelegationsPoints> => {
```
[the remark this answers is not in the document]
Comment (jeremy-babylonlabs) [reply src/app/api/getDelegationPoints.ts:31]: I think either way is fine imo, if we want to make things more clear I can split it

File: src/app/api/getDelegationPoints.ts
Comment (gbarkhatov) [inline src/app/api/getDelegationPoints.ts:1]: I feel that that this should be simplified / split. We're putting too much here

File: src/app/components/Points/DelegationPoints.tsx [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+export const DelegationPoints: React.FC<DelegationPointsProps> = ({
+  stakingTxHash,
+}) => {
+  const { delegationPoints, isLoading } = useDelegationsPoints();
+
+  const points = delegationPoints.get(stakingTxHash);
+
+  if (isLoading) {
```
Comment (gbarkhatov) [inline src/app/components/Points/DelegationPoints.tsx:16]: We need to check out the UX. For example when we switch tabs, we will the "loading", but it is slow at the moment

File: src/app/components/Points/StakerPoints.tsx [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+  publicKeyNoCoord,
+}) => {
+  const { data: stakerPoints, isLoading } = useQuery({
+    queryKey: ["stakerPoints", publicKeyNoCoord],
+    queryFn: () => getStakersPoints([publicKeyNoCoord]),
+    enabled: !!publicKeyNoCoord,
+    refetchInterval: 60000, // Refresh every minute
+  });
+
+  if (isLoading) {
```
Comment (gbarkhatov) [inline src/app/components/Points/StakerPoints.tsx:20]: Same here, though I like the "placeholder" loading

File: src/app/components/Points/DelegationPoints.tsx [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+      <div className="flex items-center justify-end gap-1">
+        <div className="h-5 w-12 animate-pulse rounded bg-gray-300 dark:bg-gray-700"></div>
+      </div>
+    );
+  }
+
+  return (
+    <div className="flex items-center justify-end gap-1">
+      <p className="whitespace-nowrap font-semibold">
+        {points !== undefined ? points : 0}
```
Comment (gbarkhatov) [inline src/app/components/Points/DelegationPoints.tsx:27]: Should we show plain 0? On mobile it might not be obvious, what this zero is: https://i.imgur.com/yu5ztXn.png

File: src/app/components/Staking/Staking.tsx
Comment (gbarkhatov) [inline src/app/components/Staking/Staking.tsx:1]: Do we just change the order of params in this file?

File: src/app/context/api/DelegationsPointsProvider.tsx
Comment (gbarkhatov) [inline src/app/context/api/DelegationsPointsProvider.tsx:1]: Okay with the current solution, think that we will switch from these context providers (to manage our global state) later on

**Review by gbarkhatov — APPROVED** [review]
@jeremy-babylonlabs I like the loading placeholder
Though each time I switch to another tab and come back, it will be loading again: https://i.imgur.com/SRp80pc.png

What do you guys think about this UX? We can make it load once per X amount of time, or leave it as it is @vitsalis @jrwbabylonlab 

---

@jeremy-babylonlabs overall really good work, I approve it from my side since the comments are minor and this implementation suits our needs at the moment

File: src/app/components/Points/DelegationPoints.tsx (hunk 20 above)
Comment (jeremy-babylonlabs) [reply src/app/components/Points/DelegationPoints.tsx:27]: should we show `-` instead? or some text representation?

Comment (gbarkhatov) [reply src/app/components/Points/DelegationPoints.tsx:27]: @vitsalis @jrwbabylonlab what do you think? I think on mobile we can do "Points: x", and on desktop just "X" since "Points" is the part of so-called "table"

**Comment by jeremy-babylonlabs** [comment]
> > @jeremy-babylonlabs I like the loading placeholder Though each time I switch to another tab and come back, it will be loading again: https://i.imgur.com/SRp80pc.png
> > What do you guys think about this UX? We can make it load once per X amount of time, or leave it as it is @vitsalis @jrwbabylonlab
> > @jeremy-babylonlabs overall really good work, I approve it from my side since the comments are minor and this implementation suits our needs at the moment
> 
> It looks wrong me to to reload when switching tabs. Does it mean i'm going to hammering the api service if i keep switching the tabs? The points are not realtime, so polling per each X minutes is fine

this is resolved now

File: src/app/api/getPoints.ts [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+  data: DelegationPoints[];
+  pagination: Pagination;
+}
+
+export interface DelegationsPoints {
+  data: DelegationPoints[];
+}
+
+export const getStakersPoints = async (
+  stakerBtcPk: string[],
```
Comment (vitsalis) [inline src/app/api/getPoints.ts:37]:
```suggestion
  stakerBtcPks: string[],
```

File: src/app/api/getPoints.ts [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+export interface PaginatedDelegationsPoints {
+  data: DelegationPoints[];
+  pagination: Pagination;
+}
+
+export interface DelegationsPoints {
+  data: DelegationPoints[];
+}
+
+export const getStakersPoints = async (
```
Comment (vitsalis) [inline src/app/api/getPoints.ts:36]: Wonder whether we should have an API for finality provider points. I guess the front-end doesn't need it, but fps might ask cc @jrwbabylonlab

File: src/app/components/Delegations/Delegation.tsx [hunk 23]
```diff
@@ -162,6 +164,11 @@ export const Delegation: React.FC<DelegationProps> = ({
             <Tooltip id={`tooltip-${stakingTxHash}`} className="tooltip-wrap" />
           </div>
         </div>
+        {isApiNormal && !isGeoBlocked && (
```
Comment (vitsalis) [inline src/app/components/Delegations/Delegation.tsx:167]: Wouldn't a staker only see this table if they connect their wallet, meaning that they are not geo-blocked?

File: src/app/components/Delegations/Delegations.tsx [hunk 24]
```diff
@@ -233,11 +273,14 @@ export const Delegations: React.FC<DelegationsProps> = ({
         </div>
       ) : (
         <>
-          <div className="hidden grid-cols-5 gap-2 px-4 lg:grid">
+          <div className="hidden grid-cols-6 gap-2 px-4 lg:grid">
             <p>Amount</p>
             <p>Inception</p>
             <p className="text-center">Transaction hash</p>
             <p className="text-center">Status</p>
+            {isApiNormal && !isGeoBlocked && (
```
Comment (vitsalis) [inline src/app/components/Delegations/Delegations.tsx:281]: Same here on the display of this table

**Review by gbarkhatov — CHANGES_REQUESTED** [review]
@jeremy-babylonlabs 
https://i.imgur.com/RmrsF5s.png
https://i.imgur.com/AEkVll6.png

File: src/app/components/Points/DelegationPoints.tsx [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+
+interface DelegationPointsProps {
+  stakingTxHash: string;
+}
+
+export const DelegationPoints: React.FC<DelegationPointsProps> = ({
+  stakingTxHash,
+}) => {
+  const { delegationPoints, isLoading } = useDelegationsPoints();
+  const isMobile = useMediaQuery("(max-width:1000px)");
```
Comment (gbarkhatov) [inline src/app/components/Points/DelegationPoints.tsx:14]: We might have issues using JavaScript for media queries
This is why all the previous solutions contained pure CSS solutions to check whether or not we are on mobile or the desktop. This is the first introduction of `useMediaQuery`

If possible, please use the tailwind breakpoints

File: src/app/components/Delegations/Delegation.tsx (hunk 23 above)
Comment (vitsalis) [reply src/app/components/Delegations/Delegation.tsx:167]: Sure

File: src/app/components/Staking/Staking.tsx [hunk 26]
```diff
@@ -1,5 +1,5 @@
 import { useQuery, useQueryClient } from "@tanstack/react-query";
-import { Transaction, networks } from "bitcoinjs-lib";
+import { networks, Transaction } from "bitcoinjs-lib";
```
[the remark this answers is not in the document]
Comment (jeremy-babylonlabs) [reply src/app/components/Staking/Staking.tsx:2]: yes

**Comment by jeremy-babylonlabs** [comment]
> @jeremy-babylonlabs https://i.imgur.com/RmrsF5s.png https://i.imgur.com/AEkVll6.png

resolved

File: src/app/api/getPoints.ts [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+  stakerBtcPks: string[],
+): Promise<StakerPoints[]> => {
+  const params: Record<string, string> = {};
+
+  params.staker_btc_pk =
```
Comment (gbarkhatov) [inline src/app/api/getPoints.ts:41]: I would add a comment there about why do we need that

**Review by gbarkhatov — APPROVED** [review]
Nice work!
